jobs/mysql_restore_job.go

```diff
@@ -0,0 +1,382 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jobs
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	xtrabackupBin       = "xtrabackup"
+	xbcloudBin          = "xbcloud"
+	xbstreamBin         = "xbstream"
+	qpressBin           = "qpress"
+	mySQLServiceName    = "mysql"
+	mySQLUserName       = "mysql"
+	mySQLGroupName      = "mysql"
+	mySQLDirectory      = "/var/lib/mysql"
+	stopTimeout         = 5 * time.Second
+	activeCheckInterval = time.Second
+)
+
+type MySQLRestoreJob struct {
+	id       string
+	timeout  time.Duration
+	l        *logrus.Entry
+	name     string
+	location BackupLocationConfig
+}
+
+// S3LocationConfig contains required properties for accessing S3 Bucket.
+type S3LocationConfig struct {
+	Endpoint     string
+	AccessKey    string
+	SecretKey    string
+	BucketName   string
+	BucketRegion string
+}
+
+// BackupLocationConfig groups all backup locations configs.
+type BackupLocationConfig struct {
+	S3Config *S3LocationConfig
+}
+
+func NewMySQLRestoreJob(id string, timeout time.Duration, name string, locationConfig BackupLocationConfig) *MySQLRestoreJob {
+	return &MySQLRestoreJob{
+		id:       id,
+		timeout:  timeout,
+		l:        logrus.WithFields(logrus.Fields{"id": id, "type": "mysql_restore"}),
+		name:     name,
+		location: locationConfig,
+	}
+
+}
+
+// ID returns job id.
+func (j *MySQLRestoreJob) ID() string {
+	return j.id
+}
+
+// Type returns job type.
+func (j *MySQLRestoreJob) Type() string {
+	return "mysql_restore"
+}
+
+// Timeouts returns job timeout.
+func (j *MySQLRestoreJob) Timeout() time.Duration {
+	return j.timeout
+}
+
+func binariesInstalled() error {
+	if _, err := exec.LookPath(xtrabackupBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xtrabackupBin)
+	}
+
+	if _, err := exec.LookPath(xbcloudBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbcloudBin)
+	}
+
+	if _, err := exec.LookPath(xbstreamBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", xbstreamBin)
+	}
+
+	if _, err := exec.LookPath(qpressBin); err != nil {
+		return errors.Wrapf(err, "lookpath: %s", qpressBin)
+	}
+
+	return nil
+}
+
+// stdout and stderr could be returned even if rerr is not nil
+func restoreMySQLFromS3(
+	ctx context.Context,
+	backupName string,
+	config *BackupLocationConfig,
+	targetDirectory string,
+) (stdout, stderr *bytes.Buffer, rerr error) {
```

Comment:
Can we eliminate named function results? Typically it's harder to understand what we return.

Reply:
I'm not using stdout and stderr variables, it's just for indicating what returned in what order.

Reply:
And `rerr` ?

Reply:
>  it's just for indicating what returned in what order.
Better to use comments for that

Reply:
rerr is used :)

Reply:
> Can we eliminate named function results?
🙂

Reply:
will `(_, _ *bytes.Buffer, rerr error)` work for you? Since rerr is used in the defer.

Reply:
rerr is left only, came up with leaving as is.

Reply:
Ok, seems that we can't eliminate)